enabling localization of variable and parameter blocks

annaxuphoto opened the pull request:
<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->

Currently, translations do not show up for variable blocks in Blockly ([ex: Artist](https://studio.code.org/projects/artist)), like `length`, `counter`, and `sides` in the `Variables` tab in Artist. Block definitions exist with variable translations, but those blocks aren't getting used in all levels. To enable translations of these blocks, we're adding variable and parameter blocks to the i18n sync-in and updating the block content on localization. The translations will end up in `variable_names.<locale>.json` and `parameter_names.<locale>.json`
## Links
[jira](https://codedotorg.atlassian.net/secure/RapidBoard.jspa?rapidView=28&modal=detail&selectedIssue=FND-1200)
<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
- jira ticket: [jira](https://codedotorg.atlassian.net/secure/RapidBoard.jspa?rapidView=28&modal=detail&selectedIssue=FND-1200)
-->

## Testing story

Ran the sync-in locally and ensured that `variable_names` and `parameter_names` were generated. Manually mocked some translations and ensured they appeared in Artist.

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

Hamms reviewed (commented):
I'd be interested to see what the actual data generated by running the sync in looks like. In particular, I wonder if it makes sense to separate out variables by level, or if we would rather have a unified set of variable names across all levels.

I'd also love to see some tests!

bethanyaconnor commented:
+1 to tests -- we have some [here](https://github.com/code-dot-org/code-dot-org/blob/7c800cf5a287a4080cb0412e58b6344d5027cf08/dashboard/test/helpers/levels_helper_test.rb#L656) as an example.

Overall the code change looks good though!

annaxuphoto commented:
![Screen Shot 2021-02-22 at 6 14 02 PM](https://user-images.githubusercontent.com/16494556/108794036-e164d700-7539-11eb-94d0-59184aca8a35.png)
![Screen Shot 2021-02-22 at 6 14 21 PM](https://user-images.githubusercontent.com/16494556/108794037-e32e9a80-7539-11eb-9365-acd786d14c49.png)

Here are some examples of the new variables in `projects.json`. There's definitely a good amount of overlap across projects and levels - most of what I'm seeing in course content is a cluster of 5 or so levels will share the same `variable_names`, and then the next cluster will introduce a new variable into those levels. It's varied enough that I'm still comfortable with separating by levels, but open to discussion.

Additionally, I've added a unit test similar to Bethany's example for variable translation.

bethanyaconnor reviewed (approved):
Excited to see this happening! I would also lean towards keeping them by level but don't feel strongly about it so will defer to whatever you decide. :)

Hamms commented:
[Here's the full diff](https://gist.github.com/Hamms/075b14f41e55c85b0f548d31f73e16a5), for reference. (note that it also includes the addition of placeholder texts; I stripped those out when performing the following analysis)

Looks like there are a total of 93 distinct new strings being added, but when breaking them up by level there are some 1500 new strings total. That's a lot of duplicates. I think it would make things a lot easier for translators if we flattened this data, like we do for block categories and progression strings.

annaxuphoto commented:
Thanks for doing that analysis, Elijah! For block categories and progressions, do we change the way the sync-out handles the output? The sync-in looks roughly the same for those strings as the others.

Hamms commented:
I don't think we do anything special for those in the sync-out, no. Note that because we don't break them up by level, the final output is quite simple: https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/config/locales/block_categories.en.yml

But note that we _do_ have some special logic in the sync in for those strings. See https://github.com/code-dot-org/code-dot-org/blob/48b686d91485c22bd9998998e64a1393446b9e8b/bin/i18n/sync-in.rb#L196-L203 for example

annaxuphoto commented:
Updated the sync-in to flatten variable strings - additionally refactored some duplicate code and raw string arguments.

Hamms commented on bin/i18n/sync-in.rb:
```diff
@@ -55,10 +55,11 @@ def get_i18n_strings(level)
 
     # authored_hints
     if level.authored_hints
+      authored_hints_string = 'authored_hints'
```
this seems unnecessary, both here and below. I would _much_ prefer we use the hash keys directly

annaxuphoto replied:
Can you explain the preference? I agree the variable doesn't improve /too/ much, but it seems like using the same raw strings in different places introduces more risk of mis-alignment.

Hamms replied:
Hiding the actual keys being used under a variable probably does reduce the risk of misalignment a bit in theory, but in practice there's no real difference between introducing a typo into `i18n_strings[authored_hints_string]` vs `i18n_strings['authored_hints']`; both will manifest as an error that will be easily noticeable. More importantly, though, it makes the code significantly more difficult to read. Interacting with hashes by passing strings and symbols as keys is standard; when I see a variable being used I expect that we're interacting with the hash programmatically rather than manually (as is the case here), and have to go searching for the variable declaration to realize that no, we're actually not doing anything programmatic here but are instead just making it look like we are.

If you did _really_ want to avoid using strings, then the right way to do it would be to declare these variables as constants in some more-globally-accessible spot in the code.

annaxuphoto replied:
Got it. I would like those to be globally accessible constants, but will count them as out of scope for this PR. Thanks for the explanation!

Hamms commented on bin/i18n/sync-in.rb:
```diff
@@ -248,23 +285,19 @@ def localize_level_content
+  write_to_yml(parameter_names, parameter_strings)
+end
+
+def write_to_yml(block_type, block_strings)
```
I don't think anything here is block-specific, is it? Should probably be something like

```suggestion
def write_to_yml(type, strings)
```

Hamms reviewed (commented):
Looks like with this approach, we still have variable names and parameter names showing up in a few places in `course_content`; my guess would be in project levels and contained levels. Probably worth going through all the places we call `get_i18n_strings` to see if there's a way to make it work more consistently

annaxuphoto commented:
Updated the sync-in to prevent variable strings from appearing in project and sublevel content. I noticed block_categories were also showing up in sublevel content, so I went ahead and removed those as well.

annaxuphoto commented:
Discussed offline with @Hamms that we also want to add project variables to the sync. In order to do so, I adjusted the level and project methods to pass the overall variable and parameter hash references as arguments, in a wrapper method that handles writing the hashes to the appropriate file as well.

Hamms reviewed (approved):
This looks great! Thanks for accommodating all my nitpicking :smile:
